File: c++/CloneGraph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        map<UndirectedGraphNode*, UndirectedGraphNode*> nodes;
        if (node == nullptr) return nullptr;
        UndirectedGraphNode* copy = new UndirectedGraphNode(node -> label);
        nodes[node] = copy;
        queue<UndirectedGraphNode*> q;
        q.push(node);
        while(!q.empty()){
            UndirectedGraphNode* cur = q.front();
            q.pop();
            //int n = cur->neighbors.size();
            for (auto neighbor :cur->neighbors ){
                if (nodes.find(neighbor)== nodes.end()){
                    UndirectedGraphNode* neigh_copy = new UndirectedGraphNode(neighbor -> label);
                    nodes[neighbor] = neigh_copy;
                    q.push(neighbor);
                }
                nodes[cur]->neighbors.push_back(nodes[neighbor]);
            }
        }
        
        return copy;
    }
};

File: python/clone_graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []


class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        clones = {}
        queue = collections.deque([])
        if node is None:
            return None
        clone = UndirectedGraphNode(node.label)
        clones[node.label] = clone
        queue.append(node)
        visited = set()
        while(queue):
            cur = queue.popleft()
            if not cur.label in visited:
                cur_clone = clones[cur.label]
                visited.add(cur.label)
                for neighbor in cur.neighbors:
                    if clones.get(neighbor.label, False) is False:
                        new_node = UndirectedGraphNode(neighbor.label)
                        clones[neighbor.label] = new_node
                    queue.append(neighbor)
                    cur_clone.neighbors.append(clones[neighbor.label])

        return clone
